### ISSUE TITLE
What's happening with container status vs isFolder()?

### ISSUE THREAD

**Issue by BobRay** [modxcms/revolution#13722]
Just trying to get a definitive answer on this for my book.

At one time the 'container' field and 'isFolder' were completely independent. Resources with no children could be containers and resources with children could be non-containers -- isFolder() was based on whether the resource had children.

This changed in a later version to be like Evolution, where the two were synchronized. Any resource with children automatically became a container when saved. If the children were removed, it automatically became a non-container. So isFolder() always matched the container status.

In 2.6.0, this seems to have reverted to having container status and isFolder() be independent.  Resources with no children can be containers and resources with children can be non-containers. 

Is this last change permanent?

**Comment by mrhaw** [comment]
> In 2.6.0, this seems to have reverted to having container status and hasChildren() be independent. Resources with no children can be containers and resources with children can be non-containers.

How it should be!!

**Comment by OptimusCrime** [comment]
Not sure if I understand. What changes are you talking about? The code [modresource.class.php#L812-L818](https://github.com/modxcms/revolution/blame/10692efe83cdd6a33571e16756c243a7ab2aadbd/core/model/modx/modresource.class.php#L812-L818) clearly indicates that the value of `hasChildren()` is calculated by the number of resources that have that resource as their parent ID. The blame also indicates that this particular function has not been changed since august of 2009

**Comment by BobRay** [comment]
Yes, hasChildren() has never changed. But I believe at one time, there was a check when saving a resource that reset isfolder both ways based on the existence of children.  This is how Evo works.

Currently, adding a second or later child will also set the parent as a container.

The current code is inconsistent imo. As far as I can tell, you can designate a resource with no children as a container and that will never change. You can also designate a resource with numerous children as a non-container and that will persist until another child is created, which could be quite a while later. 

I think that the isfolder setting should either be fully automatic, or fully manual. As it is now, sometimes it's automatic and sometimes it's not, and I doubt if anyone could guess what will happen in every case. 

My preference would be to decouple the two (iow, make it fully manual and never change isfolder unless the user changes it). It speeds up saving, allows more flexibility in styling, prevents surprising changes later on that could break external links to pages, and is easily predictable. That way, hasChildren() would report whether the resource has children (as it always has), and isfolder would be a user choice for how URLs are displayed. I'm pretty sure this was the situation when I wrote the book -- adding children did not reset isfolder.

I don't feel strongly about the point above,, but I do feel strongly that if isfolder is going to be set automatically, it should be done consistently and checked and reset whenever a resource is saved. That, BTW, would make hasChildren() redundant since checking isfolder would be faster and would always match the return from hasChildren().

**Comment by OptimusCrime** [comment]
Like I said, because `isfolder` is an attribute in the database, you can always get "cheat" the system into believing a container is not actually a container, and verse versa. 

I am not entirely sure what kind of answer you want. If it is an official one, or if you are just wondering about the current behavior. I know you are familiar with the core code, and you probability already checked the processor code, but I can answer what I can from the digging I did. Although this might not be the answer you want.

Some of the relevant processor code can be found in [/processors/resource/create.class.php#L597-L611](https://github.com/modxcms/revolution/blob/10692efe83cdd6a33571e16756c243a7ab2aadbd/core/model/modx/processors/resource/create.class.php#L597-L611). It checks a system setting called `auto_isfolder`, that has a default value of `true`. If the system setting is `true`, then it will set the parent resource to `isfolder`, regardless of how many children it already possesses. This feature was added in january 2015, according to the [blame](https://github.com/modxcms/revolution/blame/10692efe83cdd6a33571e16756c243a7ab2aadbd/core/model/modx/processors/resource/create.class.php#L604-L606), in [2.3.3-pl](https://github.com/modxcms/revolution/releases/tag/v2.3.3-pl) as a part of commit 5041463e920d815c1274122a6d8c3cc8ea280751. The description of system setting `auto_isfolder` is simply: "If set to yes, container property will be changed automatically.".  Similarly, for sorting, if the same system setting is true, a check will be done when removing a child. If the current resource have no children left, the `isfolder` attribute is set to `false`. [Relevant code in commit](https://github.com/modxcms/revolution/commit/5041463e920d815c1274122a6d8c3cc8ea280751#diff-2f26d694a02a06bdc2ef79bc952125f9R187).

The issue was discussed in several issues. Some of these includes: #5812, #9401, #9267 and #3508. Some noticable comments from these issues includes:

@opengeek [comment](https://github.com/modxcms/revolution/issues/9267#issuecomment-33816378) on #9267:
>There is a big difference between isfolder and hasChildren in MODX Resources. Changing this field without user intention will also change the URL for the Resource, potentially without the user's knowledge. I'm of the opinion that this should only be changed by direct user action since it does affect routing to the Resource.

@theboxer [comment](https://github.com/modxcms/revolution/issues/9401#issuecomment-45141240) on #9401:
>If the resource is a document (and it has children), that the resource has a document icon.
>If the resource is marked as container, than the resource has a container icon.

This was what I could dig up from the older history. I think @opengeek described it best. They are different functions that does different things. Also note that `modResource::hasChildren` is not used _a single place_ in the entire MODX core. This functional is just to make it easier for the developers to do common routines, like checking if a resource has any children.

Apologies if this was not the type of answer you were looking for.

**Comment by BobRay** [comment]
Thanks for all the research. It helped me focus on my issue and realize that I've confused things by mentioning hasChildren(), which is beside the point.  

Originally, I just wanted an official answer, but after seeing your research and doing some testing, I think there's a serious problem with the current code.  I agree with this statement on the isfolder field by @opengeek (emphasis mine):

> Changing this field without user intention will also change the URL for the Resource, potentially without the user's knowledge. I'm of the opinion that this **should only be changed by direct user action** since it does affect routing to the Resource.

Consider this scenario:  A user creates a parent resource and several children under it, then unchecks the container checkbox for the parent and saves it. Much later, the user adds another child and the parent (which the user is not editing) becomes a container. External links to the parent (and potentially all children) are now broken. The user has no idea that this has happened, and if alerted, may not know how to fix it. In addition, the user's fix may be temporary if another child is added later.

The problem is exacerbated by the fact that auto_isfolder is on by default.

I see only two ways out of this.

1. Following @opengeek's path -- remove any code that modifies isfolder after the fact and remove the auto_isfolder System Setting.

2. Add code to test and set isfolder for the current resource and its parent on every resource save or update if auto_isfolder is on.  Note that extras that add or remove resources without calling a processor would have to be rewritten to check auto_isfolder and set the isfolder field correctly unless the test is in the modResource->save(), xPDOObject>remove() methods and wherever the code is that sets the 'deleted' field.

 My vote would be for the first option.

**Comment by OptimusCrime** [comment]
I disagree with the two solutions, as well as this being a problem.

The `isfolder` property is designed in a way that makes sense for around 99% of all use cases for friendly URLs. The _only_ thing this property is used for, [as far as I can tell](https://github.com/modxcms/revolution/blob/719c29231279b44cd9a6f5cd5a49f181a44a63dd/core/model/modx/modresource.class.php#L895-L897), is to set the extension value. The description for this behavior is explained like this: "set extension to container suffix if Resource is a folder, HTML content type, and the container suffix is set".

In other words, the property is only used to construct URLs like `path/to/resource.html`, instead of havig `path.html/to.html/resource.html`. Across the WWW, you would almost _never_ want to keep the extension for the parents of a resource. They are usually only displayed for the last child. How would your suggestion handle this?

I would also like to add that the current behavior has been the default since the beginning. We have always constructed URLs like this. The system setting was added just to fix the rare edge cases. The solution you propose are breaking changes and could not, by design, go into 2.x.

You provide an example, but I fail to see any real world value of this. When would this ever be necessary to do? Note that MODX currently supports both these URLs `path/to/` and `path/to` where the resource with alias `to` is a container with children. In other words, it really does not play a huge role in the generation of URLs, other than the example I provided above.

The discussion in issue #9401 was concluded with the same arguments I provided here. I would also like to add that across the thousands of sites that run MODX, with a version history going back to the beginning, very few issues about `isfolder` have been raised. Except the bugs/discussions linked in my previous comment, I can not see anyone fundamentally disagreeing with the implementation like it is today.

But if you have a good alternative to the problems with alias paths I illustrated above, I am all ears.

**Comment by opengeek** [comment]
@OptimusCrime This has absolutely nothing to do with alias paths. This is about the extension added to the actual resource. It would never be added to parent paths. I'm a little confused on how this got into the discussion.

**Comment by OptimusCrime** [comment]
@opengeek Really? @theboxer stated the same in [his comment](https://github.com/modxcms/revolution/issues/9401#issuecomment-45143638). That was also how I interpreted the code.

**Comment by opengeek** [comment]
I think he must have been confused. Parent aliases do not get extensions when building URIs with use_alias_paths enabled. Never have, and never will.

**Comment by opengeek** [comment]
This is only regarding the extension that is added to the final path segment in a URI.

**Comment by OptimusCrime** [comment]
So this behavior is just when creating the URIs for resources that are parents? It decides if it should be `path/to/parent.html` or `path/to/parent/`?

**Comment by opengeek** [comment]
Correct. The is_folder attribute determines if it gets the container_suffix or the extension for the specified content_type. My opinion has not changed. Specifically, is_folder should _only_ be used to determine how the URI is constructed (i.e. if it behaves like a folder on a traditional web server) and it should not be automatically changed on any type of edit, or if children are added to or removed from it. I believe changing the default of auto_isfolder to false would be prudent at the very least.

**Comment by BobRay** [comment]
This has definitely changed more than once. Prior to 2.0.0 it was automatic based on the existence of children (a la Evo). 

In the changelog, these happened in 2.0.0
  - [#MODX-1226] Removed modResource->checkChildren() method; isfolder should not be set based on presence/absence of children.
- [#MODX-535] Removed automatic setting of isfolder based on presence or absence of children.

At present, it's partially automatic based on the existence of children but not reliably so. if isfolder is turned off, it will stay off until another child is created. That's bad, imo. And even with auto_isfolder on, a resource with no children can have isfolder set to No permanently.

All that said, after thinking some more about it, if isfolder is never set automatically,  there's a problem that's the opposite of my scenario above -- the user forgets to set the container checkbox, sets it later and external links break.

I'm thinking that if auto_isfolder is on, we should make sure isfolder is always based on the existence of children. I don't have a problem with auto_isfolder being on by default as long as it's reliable.  As OptimusCrime points out, it would be a rare case where someone would want to use isfolder in some other way. 

With auto_isfolder on, the Container checkbox should probably be read-only or hidden.

**Comment by OptimusCrime** [comment]
The only way to ensure this is by doing running a SQL query.

**Comment by mrhaw** [comment]
I think it's perfect the way it is now and no change is needed. These options makes it possible to build
any type of WayFinder menu possible on the fly. Right now the auto_isfolder is more of a helper than a rule or restriction. 

I don't see it being a problem that a "_resource with no children can have isfolder set to No permanently._" 

Maybe a whole new system setting is what is requested here? "override_isfolder" and only depend on site tree hierarchy and hasChildren...

**Comment by BobRay** [comment]
Sorry, I said that wrong:  Currently, a resource with no children can have isfolder set to **YES** permanently.

As I said above, my position is that, "if auto_isfolder is on, we should make sure isfolder is always based on the existence of children." It is not now.

Restoring the fixParents() code in the create and update processors (I know it's there in V 2.4.4) would be one way, but it could also be done in the resource->save() method.
